content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

```diff
@@ -0,0 +1,581 @@
+
+This is also the part of the parser where we look at precedence, which also confused me.
+
+If I were to spend more time on this book, I would focus on section 2.7 - How Pratt Parsing Works because it's clear to me as I write this that I don't fully comprehend what is happening.
+
+#### parseBoolean
+
+Inside our `parser/parser.go` file, we add a helper function:
+
+```go
+func (p *Parser) parseBoolean() ast.Expression {
+  return &ast.Boolean{
+    Token: p.curToken,
+    Value: p.curTokenIs(token.TRUE)
+    }
+}
+```
+
+I wrote a question mark next to the `token.TRUE` because I was confused on why this works even when parsing a `false` boolean. Again, this is probably do to my lack of understanding here and another area I would focus on again.
+
+#### grouped expressions
+
+Some how, we can group expressions like `(5 + 5) * 2` to influence the parsing to our liking. We add a function called `parseGroupedExpression` and it "just works." I couldn't tell you why though. Here's the function in case you can figure it out:
+
+```go
+func (p *Parser) parseGroupedExpression() ast.Expression {
+  p.nextToken()
+  exp := p.parseExpression(LOWEST) // this LOWEST thing I don't get
+
+  if !p.expectPeek(token.RPAREN) {
+    return nil
+  }
+
+  return exp
+}
+```
+
+#### block statements
+
+These are a series of statements which are surrounded by `{` and `}`.
+
+#### infix parse functions
+
+I wrote a note later where he says,
+
+> Yes, we need to register an `infixParseFn` for `token.LPAREN`. This way we parse the expression that is the function (either an identifier, or a function literal), then check for an `infixParseFn` associated with `token.LPAREN` and call it with the already parsed expression as argument. And in this `infixParseFn` we can then parse the argument list.
+>
+> -Ball 122
+
+I'm blocking myself here. In my mind, I think "infix === things like +, -, \*, /" so it feels strange here to use an LPAREN or "(" as an infix, but I guess infix can also mean in between two other characters. Example: `add(5, 5)` where the `(` is in between the "d" and the "5". Still, future me should revisit this.
+
+#### serialization
+
+Thorsten mentions this,
+
+> What we want is an AST that (serialized as a string)...
+>
+> -Ball 89
+
+And I thought, "Do I really know what this means?" And the answer is no, no I do not. Wikipedia [explains it](https://en.wikipedia.org/wiki/Serialization) well. I'll summarize it as taking something, translating it into a data structure or object and storing it for later use (possibly restructuring it).
```

Comment:
The colloquial term would be "dumping" 😉 As in "What we want is an AST that (dumped as a string)..."

Or "converting".

Reply:
ahhh! converting helps! 😄